North-Seattle-College/ad440-winter2022-thursday-repo
# 42 sagemaker automation script improvements

### Closes issue #42 

### Overview
- This pull request renames and updates two files: `deploySagemakerNBI.py` and `sagemakerStackTemplate.json`
- The changes allow a user to execute the SageMaker automation script with command line arguments for initials (used to create stack and notebook names) and AWS region for deployment with no additional interaction needed.
- The notebook instance name contains the provided initials for easy identification.
- This also fixes an issue where students could not deploy when CloudFormation tried to create a new Execution Role (prohibited in our account) by using an existing Role.

### Pre-requisites
- Python Boto3 installed (python -m pip install boto3)
- AWS CLI and proper basic configuration

### To Test
1. Open branch/download the scripts: `deploySagemakerNBI.py` and `sagemakerStackTemplate.json`
2. Run deploySagemakerNBI.py with the two required string arguments: `--initials` and `--region`. Example: `python deploySagemakerNBI.py --initials nwc --region us-west-2`
3. The script should provide a confirmation message that stack creation has started and then run for some time before returning a successful creation message.
4. Verify that your stack was created by going to CloudFormation in the AWS Console. 
![stack_create](https://user-images.githubusercontent.com/45498644/153989228-b2734e0f-8e41-4ccd-b4a0-78fa2429e6a3.png)
5. Verify that your Sagemaker notebook was created by going to SageMaker in the AWS Console and selecting Notebook > Notebook instances in the left menu.
![notebook](https://user-images.githubusercontent.com/45498644/153989547-4506ee2e-0fcf-4b62-83b0-3206338cd08a.png)
6. IMPORTANT: Delete your stack from the AWS Console to limit charges.

### Time Spent
| Activity | Estimate | Actual Time Spent |
| ------ | ------ | ------ |
| Research command line parameters in Python | 1 hours | 1 hours |
| Develop code improvements | 2 hours | 3 hours |
| Testing | 1 hour | 30 mins |
| Documentation | 30 mins | 30 mins |
| Pull request and wrap-up | 30 mins | 30 mins |
